google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkDevelopmentDetails.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.sdk;
+
+import com.google.cloud.tools.appengine.api.whitelist.AppEngineJreWhitelist;
+import com.intellij.execution.configurations.ParametersList;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.util.text.StringUtil;
+import gnu.trove.THashMap;
+import gnu.trove.THashSet;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CloudSdkDevelopmentDetails {
```

Comment:
javadoc; Also the name sounds a little strange, maybe something like `CloudSdkInternals`?

Also, how about giving this a static factory method?

Reply:
I like `CloudSdkInternals`, great name, thanks. I thought about using a static instance but then it operates on `CloudSdkService` instance and might be a bit different for different implementations, so we'd have to maintain a map of instances per `CloudSdkService` type, which can become too much of a caching for a simple helper class. Plus, it also makes it easier to test this way. We can still assume it operates on default `CloudSdkService` only of course. WDYT?

Reply:
when you say 

> we'd have to maintain a map of instances per CloudSdkService type

what are the different types that you mean? Don't we always have a single cloud sdk?

Reply:
If you are saying that it depends on some state of CloudSdkService, could you add a method in CloudSdkService to return a Validator of itself?

Reply:
and perhaps the validator being an inner class of CloudSdkService so that it can have access to whatever it needs (I assume its just the sdk path)?

Reply:
`CloudSdkValidator` is not always working with `CloudSdkService`, it's also used to check for validity of paths in progress, i.e. in configuration panel when the changes are not applied yet, so technically they are different and not that strongly coupled. The idea was to minimize the interface to allow it to be easily implemented and changed without moving around all of the details of validation or internals which are slightly different. They are close but feels like serve different needs.

Reply:
ok. sounds like I need to take a closer look then. sounds good for now.